### ISSUE TITLE
Line continuation not working properly?

### ISSUE THREAD

**Issue by jakwings** [fish-shell/fish-shell#2897]
I'm designing a shell scripting language, so this is nitpicking. I just compare the behavior of fish to bash's.

fish, version 2.2.0-722-g411d573

Before every test, run `set -geu PATH` and `set -geu fish_user_paths` in order to use built-in functions only. (**Update: oh, when PATH is empty, it defaults to "/usr/bin:/bin"**)

Please save every test into separate files and manually source them.

*   Test 1:

    ```fish
    rand\
    om
    ```

    prints a random number

*   Test 2:

    ```fish
    buil\
    tin random
    ```

    prints nothing

*   Test 3:

    ```fish
    true
    a\
    nd echo ok
    ```

    prints nothing

*   Test 4:

    ```fish
    if true; e\
    nd
    ```

    Error: `Missing end to balance this if statement`

*   Test 5:

    ```fish
    i\
    f true; echo ok; end
    ```

    Error: `'end' outside of a block`

*   Test 6:

    ```fish
    if\
     true; end
    ```

    Error: `'end' outside of a block`

*   Test 7:

    ```fish
    if true; end\
    ;
    ```

    Error: `Missing end to balance this if statement`

Some other keywords are also tested, and it seems that only `begin/if/while/switch/function/end` (well, maybe all buitins) do not support line continuation.

**Comment by faho** [comment]
How would you expect it to work?

Test 1 and 2 are of course clearly inconsistent, but I'd expect _all_ of these to error out. I'm not sure why I'd expect that since I'm okay with line continuations inside a "string", e.g.:

```fish
echo "Hello \
World"
```

but not inside a command word like you've demonstrated here.

**Comment by faho** [comment]
For the record, bash allows these.

**Comment by jakwings** [comment]
Yes, playing these with bash is pretty fun.

Disallowing line continuation in the command word/path is acceptable to me.

**Comment by jakwings** [comment]
Just added two similar tests. And taken these cases into consideration, I think a more consistent behavior like bash's is more plausible and easier to implement:

```fish
if\
 something interesting
end

if\
\
 something interesting
end

i\
\
f something interesting
end
```

**Comment by krader1961** [comment]
It is unlikely that someone designed and coded the bash lexical scanner and tokenizer to deliberately allow splitting tokens like `if` across multiple lines using a backslash. More likely it's an artifact of the implementation; i.e., unintentional behavior. Do we really want to change the fish parser just to make it possible to write obfuscated scripts?

FWIW, backslashes are already allowed and work as expected for more reasonable cases. For example,

```
if \
true
    echo yes
end
```

**Comment by krader1961** [comment]
On the other hand I do agree that constructions like

```
true
a\
nd echo yes
```

Shouldn't fail silently. It should either be equivalent to

```
true
and echo yes
```

Or it should emit an explicit error message.

**Comment by jakwings** [comment]
I think I've found out the cause of the problem. It is in [src/parse_tree.cpp](https://github.com/fish-shell/fish-shell/blob/4ff8e6e7818747d805ac44be0a36aedef4deaa5c/src/parse_tree.cpp#L1258). The line-continuations (`\\\n`) are still in the token text before the cleaness check. (see my pull request #2916)

It is even a surprise to see `"if"` (quoted) is also a keyword. Any good reason to keep this feature?

**Comment by ridiculousfish** [comment]
Nice work. I implemented quoted keywords in response to #1398 and I think it's the right thing to do. Quoting should not be a "back door" for running an external command.

**Comment by krader1961** [comment]
Ugh! This is going to require more changes to fish_indent for it to produce canonical output. But since this change addresses the problems documented in issue #2897 and is consistent with the earlier change made by @ridiculousfish in issue #1398 I'm going to merge it. I'm not entirely comfortable with the is_keyword_char() name since backslashes aren't really valid keyword chars but I can't think of a cleaner way to express what your change is doing within the context of the existing code.

Also, a big thank you, @jakwings, for the unit tests. Far too few changes to this project include relevant tests to help avoid regressions.
